lib/Runtime/Library/JavascriptString.cpp

```diff
@@ -3885,13 +3894,96 @@ namespace Js
+
+        Assert(leftString->GetLength() == rightString->GetLength());
+        Assert(wmemcmp(lstr, rstr, leftString->GetLength()) == 0);
+
+        bool leftIsPropertyString = VirtualTableInfo<Js::PropertyString>::HasVirtualTable(leftString);
+        bool leftIsLiteralStringWithPropertyStringPtr = VirtualTableInfo<Js::LiteralStringWithPropertyStringPtr>::HasVirtualTable(leftString);
+
+        // if one of the strings is a LiteralStringWithPropertyStringPtr and the other is PropertyString
+        // check if we could SetPropertyString if it wasn't set.
+        if (leftIsPropertyString)
+        {
+            rightString->SetBuffer(lstr);
+            if (VirtualTableInfo<Js::LiteralStringWithPropertyStringPtr>::HasVirtualTable(rightString))
+            {
+                Js::LiteralStringWithPropertyStringPtr * rightStringWithPtr = (Js::LiteralStringWithPropertyStringPtr*)rightString;
+
+                if (rightStringWithPtr->GetPropertyString() == nullptr)
+                {
+                    rightStringWithPtr->SetPropertyString((Js::PropertyString*)leftString);
+                }
+            }
+            return;
+        }
+
+        if (leftIsLiteralStringWithPropertyStringPtr)
+        {
+            Js::LiteralStringWithPropertyStringPtr * leftStringWithPtr = (Js::LiteralStringWithPropertyStringPtr*)leftString;
+            if (VirtualTableInfo<Js::PropertyString>::HasVirtualTable(rightString))
+            {
+                leftString->SetBuffer(rstr);
+                if (leftStringWithPtr->GetPropertyString() == nullptr)
+                {
+                    leftStringWithPtr->SetPropertyString((Js::PropertyString*)rightString);
+                }
+                return;
+            }
+
+            if (VirtualTableInfo<Js::LiteralStringWithPropertyStringPtr>::HasVirtualTable(rightString))
+            {
+                Js::LiteralStringWithPropertyStringPtr * rightStringWithPtr = (Js::LiteralStringWithPropertyStringPtr*)rightString;
+
+                if (leftStringWithPtr->GetPropertyString() == nullptr && rightStringWithPtr->GetPropertyString() != nullptr)
+                {
+                    leftStringWithPtr->SetPropertyString(rightStringWithPtr->GetPropertyString());
+                    leftString->SetBuffer(rstr);
+                    return;
+                }
+
+                if (rightStringWithPtr->GetPropertyString() == nullptr && leftStringWithPtr->GetPropertyString() != nullptr)
+                {
+                    rightStringWithPtr->SetPropertyString(leftStringWithPtr->GetPropertyString());
+                    rightString->SetBuffer(lstr);
+                    return;
+                }
+            }
+        }
+
+        // try to use the buffer from PropertyString
+        if (leftIsPropertyString || leftIsLiteralStringWithPropertyStringPtr)
+        {
+            rightString->SetBuffer(lstr);
+        }
+        else
+        {
+            leftString->SetBuffer(rstr);
```

Comment:
we need to be very careful about this, if right is a PropertyString, the buffer is an interior pointer